velox/functions/prestosql/tests/ArrayNormalizeTest.cpp

```diff
@@ -0,0 +1,202 @@
+    T sum = 0;
+    for (const auto& item : values) {
+      sum += pow(abs(item), p);
+    }
+
+    T pNorm = pow(sum, 1.0 / p);
+
+    std::vector<T> res;
+    for (T value : values) {
+      res.push_back(value / pNorm);
+    }
+
+    return res;
+  }
+
+  template <typename T>
+  void testArrayWithElementsEqualZero() {
+    auto input = makeArrayVector<T>({{0.0, -0.0, 0.0}});
+    auto p = makeConstant<T>(2.0, input->size());
+    testExpr(input, {input, p});
+  }
+
+  template <typename T>
+  void testArrayWithPLessThanZero() {
+    auto input = makeArrayVector<T>({{1.0, 2.0, 3.0}});
+    auto p = makeConstant<T>(-1.0, input->size());
+    EXPECT_THROW(testExpr(input, {input, p}), facebook::velox::VeloxUserError);
+  }
+
+  template <typename T>
+  void testArrayWithPEqualZero() {
+    auto vector = makeArrayVector<T>({{1.0, -2.0, 3.0}});
+    auto p = makeConstant<T>(0.0, vector->size());
+    testExpr(vector, {vector, p});
+  }
+
+  template <typename T>
+  void testArrayWithPLessThanOne() {
+    T pValue = 0.5;
+    std::vector<T> values = {1.0, -4.0, 9.0};
+    std::vector<T> expected = normalize(values, pValue);
+
+    auto input = makeArrayVector<T>({values});
+    auto p = makeConstant<T>(pValue, input->size());
+    testExpr(makeArrayVector<T>({expected}), {input, p});
+  }
+
+  template <typename T>
+  void testArrayWithPEqualOne() {
+    T pValue = 1.0;
+    std::vector<T> values = {1, -2, 3};
+    std::vector<T> expected = normalize(values, pValue);
+
+    auto input = makeArrayVector<T>({values});
+    auto p = makeConstant<T>(pValue, input->size());
+    testExpr(makeArrayVector<T>({expected}), {input, p});
+  }
+
+  template <typename T>
+  void testArrayWithTypeLimits() {
+    auto input = makeArrayVector<T>({
+        {},
+        {0},
+        {1},
+        {std::numeric_limits<T>::infinity()},
+        {std::numeric_limits<T>::quiet_NaN()},
+        {std::numeric_limits<T>::lowest(), -1.0},
+        {std::numeric_limits<T>::max(), 1.0},
+    });
+    auto p = makeConstant<T>(2.0, input->size());
+    auto expected = makeArrayVector<T>(
+        {{},
+         {0},
+         {1},
+         {std::numeric_limits<T>::quiet_NaN()},
+         {std::numeric_limits<T>::quiet_NaN()},
+         // For the cases below, pNorm is calculated as Infinity, resulting in
+         // zero values. The result matches the Presto Java version.
+         {-0.0, -0.0},
+         {0.0, 0.0}});
+    testExpr(expected, {input, p});
+  }
+
+  template <typename T>
+  void testArrayWithNullValues() {
+    auto input = makeNullableArrayVector<T>(
+        {{{1, 2, std::nullopt}}, {{std::nullopt}}, std::nullopt});
+    auto p = makeConstant<T>(2.0, input->size());
+    auto expected =
+        makeNullableArrayVector<T>({std::nullopt, std::nullopt, std::nullopt});
+    testExpr(expected, {input, p});
+  }
+
+  template <typename T>
+  void testArrayWithDifferentValues() {
+    T pValue = 4.0;
+
+    std::vector<T> arr1 = {1.0};
```

Comment:
variable names like this are anti-pattern; consider inlining

```
auto input = makeArrayVector<T>({
      {1.0},
      {-1.0, -1.0},
      ...
   });
```

Reply:
I use those arrays for the input and result. I cannot inline them unless I duplicate the code which I was asked not to do in the earlier rounds of review. I will rename the variables instead.